examples/exex/rollup/src/db.rs

```diff
@@ -0,0 +1,231 @@
+};
+use reth_interfaces::provider::ProviderResult;
+use reth_primitives::{
+    revm_primitives::{AccountInfo, Bytecode},
+    trie::AccountProof,
+    Account, Address, BlockNumber, BlockWithSenders, Bytes, Receipt, Receipts, StorageKey,
+    StorageValue, B256, U256,
+};
+use reth_provider::{
+    AccountReader, BlockHashReader, BundleStateWithReceipts, OriginalValuesKnown, ProviderError,
+    StateProvider, StateRootProvider,
+};
+use reth_trie::updates::TrieUpdates;
+use rusqlite::Connection;
+
+pub struct Database {
+    connection: Arc<Mutex<Connection>>,
+}
+
+impl Database {
+    pub fn new(connection: Connection) -> eyre::Result<Self> {
+        let database = Self { connection: Arc::new(Mutex::new(connection)) };
+        database.create_tables()?;
+        Ok(database)
+    }
+
+    fn connection(&self) -> MutexGuard<'_, Connection> {
+        self.connection.lock().expect("failed to acquire database lock")
+    }
+
+    fn create_tables(&self) -> eyre::Result<()> {
+        self.connection().execute_batch(
+            "CREATE TABLE IF NOT EXISTS account (
+                id      INTEGER PRIMARY KEY,
+                address TEXT UNIQUE,
+                data    TEXT
+            );
+            CREATE TABLE IF NOT EXISTS bytecode (
+                id   INTEGER PRIMARY KEY,
+                hash TEXT UNIQUE,
+                data TEXT
+            );
+            CREATE TABLE IF NOT EXISTS storage (
+                id      INTEGER PRIMARY KEY,
+                address TEXT,
+                key     TEXT,
+                data    TEXT,
+                UNIQUE (address, key)
+            );
+            CREATE TABLE IF NOT EXISTS block (
+                id     INTEGER PRIMARY KEY,
+                number TEXT UNIQUE,
+                hash   TEXT
+            );",
+        )?;
+        Ok(())
+    }
+
+    pub fn insert_block(
+        &mut self,
+        block: &BlockWithSenders,
+        bundle: BundleState,
+    ) -> eyre::Result<()> {
+        let mut connection = self.connection();
+        let tx = connection.transaction()?;
+
+        tx.execute(
+            "INSERT INTO block (number, hash) VALUES (?, ?)",
+            (block.header.number.to_string(), block.hash_slow().to_string()),
+        )?;
+
+        let StateChangeset { accounts, storage, contracts } =
+            bundle.into_plain_state(OriginalValuesKnown::Yes);
+
+        for (address, account) in accounts {
+            if let Some(account) = account {
+                tx.execute(
+                    "INSERT INTO account (address, data) VALUES (?, ?) ON CONFLICT(address) DO UPDATE SET data = excluded.data",
+                    (address.to_string(), serde_json::to_string(&account)?),
+                )?;
+            } else {
+                tx.execute("DELETE FROM account WHERE address = ?", (address.to_string(),))?;
+            }
+        }
+
+        for PlainStorageChangeset { address, wipe_storage, storage } in storage {
+            if wipe_storage {
```

Comment:
you still need to write the contents of `storage` if `wipe_storage == true`, you just need to wipe it before inserting the new values, see eg https://github.com/paradigmxyz/reth/blob/c659e28aa0e3ea5e83341b8a9eead0f0c89c296a/crates/storage/provider/src/bundle_state/state_changes.rs#L51-L79